TestTSOFollowerProxy is flaky

## Flaky Test
### Which jobs are failing
```text
panic: test timed out after 5m0s
        	            	github.com/tikv/pd/tests/integrations/client_test.TestTSOFollowerProxy
        	            				/home/prow/go/src/github.com/tikv/pd/tests/integrations/client/client_test.go:276
        	Error:      	Received unexpected error:
```
### CI link
- https://prow.tidb.net/view/gs/prow-tidb-logs/pr-logs/pull/tikv_pd/10275/pull-unit-test-next-gen-3/2031622997105709056
### Reason for failure (if possible)

### Anything else
- triage-key: `testtsofollowerproxy`
- test: `TestTSOFollowerProxy`

--- a/tests/cluster.go
+++ b/tests/cluster.go
@@ -88,7 +88,15 @@ const (
 	startServersRetryRecreate
 )
 
-func classifyStartServersError(err error) startServersRetryAction {
+func shouldRetryCurrentServers(err error) bool {
+	if err == nil {
+		return false
+	}
+	errMsg := err.Error()
+	return strings.Contains(errMsg, "ErrCancelStartEtcd") || strings.Contains(errMsg, "ErrStartEtcd")
+}
+
+func classifyInitialServersError(err error) startServersRetryAction {
 	if err == nil {
 		return startServersNoRetry
 	}
@@ -96,7 +104,7 @@ func classifyStartServersError(err error) startServersRetryAction {
 	switch {
 	case strings.Contains(errMsg, "address already in use") || strings.Contains(errMsg, "Etcd cluster ID mismatch"):
 		return startServersRetryRecreate
-	case strings.Contains(errMsg, "ErrCancelStartEtcd") || strings.Contains(errMsg, "ErrStartEtcd"):
+	case shouldRetryCurrentServers(err):
 		return startServersRetryCurrent
 	default:
 		return startServersNoRetry
@@ -712,7 +720,7 @@ func (c *TestCluster) runInitialServersWithRetry(maxRetries int) error {
 			return nil
 		}
 
-		switch classifyStartServersError(lastErr) {
+		switch classifyInitialServersError(lastErr) {
 		case startServersRetryRecreate:
 			// `Etcd cluster ID mismatch` can happen when the allocated peer URL happens to
 			// connect to another test's etcd cluster (port reuse across concurrent `go test`
@@ -786,7 +794,7 @@ func RunServersWithRetry(servers []*TestServer, maxRetries int) error {
 			return nil
 		}
 
-		if classifyStartServersError(lastErr) == startServersRetryCurrent {
+		if shouldRetryCurrentServers(lastErr) {
 			log.Warn("etcd start failed, will retry", zap.Error(lastErr))
 			// Stop any partially started servers before retrying
 			for _, s := range servers {

--- a/tests/cluster_test.go
+++ b/tests/cluster_test.go
@@ -29,14 +29,28 @@ func TestMain(m *testing.M) {
 	goleak.VerifyTestMain(m, testutil.LeakOptions...)
 }
 
-func TestClassifyStartServersError(t *testing.T) {
+func TestShouldRetryCurrentServers(t *testing.T) {
 	t.Parallel()
 
 	re := require.New(t)
-	re.Equal(startServersRetryCurrent, classifyStartServersError(errors.New("[PD:server:ErrCancelStartEtcd]etcd start canceled")))
-	re.Equal(startServersRetryCurrent, classifyStartServersError(errors.New("[PD:server:ErrStartEtcd]start etcd failed")))
-	re.Equal(startServersRetryRecreate, classifyStartServersError(errors.New("listen tcp 127.0.0.1:2379: bind: address already in use")))
-	re.Equal(startServersRetryRecreate, classifyStartServersError(errors.New("Etcd cluster ID mismatch")))
-	re.Equal(startServersNoRetry, classifyStartServersError(errors.New("some other error")))
-	re.Equal(startServersNoRetry, classifyStartServersError(nil))
+	re.True(shouldRetryCurrentServers(errors.New("[PD:server:ErrCancelStartEtcd]etcd start canceled")))
+	re.True(shouldRetryCurrentServers(errors.New("[PD:etcd:ErrStartEtcd]start etcd failed")))
+	re.True(shouldRetryCurrentServers(errors.New("[PD:etcd:ErrStartEtcd]start etcd failed: listen tcp 127.0.0.1:2379: bind: address already in use")))
+	re.False(shouldRetryCurrentServers(errors.New("listen tcp 127.0.0.1:2379: bind: address already in use")))
+	re.False(shouldRetryCurrentServers(errors.New("Etcd cluster ID mismatch")))
+	re.False(shouldRetryCurrentServers(errors.New("some other error")))
+	re.False(shouldRetryCurrentServers(nil))
+}
+
+func TestClassifyInitialServersError(t *testing.T) {
+	t.Parallel()
+
+	re := require.New(t)
+	re.Equal(startServersRetryCurrent, classifyInitialServersError(errors.New("[PD:server:ErrCancelStartEtcd]etcd start canceled")))
+	re.Equal(startServersRetryCurrent, classifyInitialServersError(errors.New("[PD:etcd:ErrStartEtcd]start etcd failed")))
+	re.Equal(startServersRetryRecreate, classifyInitialServersError(errors.New("[PD:etcd:ErrStartEtcd]start etcd failed: listen tcp 127.0.0.1:2379: bind: address already in use")))
+	re.Equal(startServersRetryRecreate, classifyInitialServersError(errors.New("listen tcp 127.0.0.1:2379: bind: address already in use")))
+	re.Equal(startServersRetryRecreate, classifyInitialServersError(errors.New("Etcd cluster ID mismatch")))
+	re.Equal(startServersNoRetry, classifyInitialServersError(errors.New("some other error")))
+	re.Equal(startServersNoRetry, classifyInitialServersError(nil))
 }

